### PR TITLE
OTLP LogExporter to enable ParseStateValues by default

### PR DESCRIPTION
**Pull request by cijothomas**
See https://github.com/open-telemetry/opentelemetry-dotnet/pull/3183#issuecomment-1099728854

### CODE REVIEW

**Review by reyang — APPROVED** [review]
LGTM.

**Comment by cijothomas** [comment]
@alanwest @CodeBlanch Please share your thoughts/comments. I think this is a reasonable way to proceed. (will do similar change for ConsoleExporter as well), by letting individual Exporters to make it easy for their users to configure ParseStateValues. If an Exporter wants to still do its own parsing of "State", its totally legal, as `State` was part of LogRecord, before `StateValues`.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol.Logs/OtlpLogExporterHelperExtensions.cs [hunk 1]
```diff
@@ -42,7 +42,7 @@ private static OpenTelemetryLoggerOptions AddOtlpExporter(OpenTelemetryLoggerOpt
         {
             configure?.Invoke(exporterOptions);
             var otlpExporter = new OtlpLogExporter(exporterOptions);
-
+            loggerOptions.ParseStateValues = true;
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol.Logs/OtlpLogExporterHelperExtensions.cs:45]: Might be useful to have a comment here explaining why it is being done?

**Review by CodeBlanch — APPROVED** [review]
Feels a bit clunky but I don't have a better solution 😄 

Should we add more to the comments here calling out that exporters may automatically enable `ParseStateValues` if they support the feature?

https://github.com/open-telemetry/opentelemetry-dotnet/blob/50baa250806301304333007a373aeb6451fc5a3a/src/OpenTelemetry/Logs/OpenTelemetryLoggerOptions.cs#L42-L51

**Comment by alanwest** [comment]
I agree feels a bit clunky for AddOltpExporter to have the side effect of changing an option - though I also agree that `ParseStateValues=true` seems like the most reasonable default.

Why not the proposal from this comment? Are we considering it inefficient because prevents state from being garbage collected? Or some other reason? https://github.com/open-telemetry/opentelemetry-dotnet/pull/3183#issuecomment-1104327969.

> If we consider the default (ParseStateValues = false) as a bad default, we could change the default to make it true. To avoid breaking existing users, the SDK can continue to populate State as well. That way, exporters need to only deal with the parse state values, and if any exporters were previously relying on the State, they will be unaffected as well...

---

Another alternative may be to make the setting `bool? ParseStateValues`. Default is null - i.e., unspecified. Makes it clear that the actual default depends on the exporter you've configured. This is similar to what I did with `PeriodicExportingMetricReaderOptions`.

https://github.com/open-telemetry/opentelemetry-dotnet/blob/a709cfd2a993fef5dedf8beb73b128ff2d03b896/src/OpenTelemetry/Metrics/PeriodicExportingMetricReaderOptions.cs#L19-L26

**Review by alanwest — APPROVED** [review]
Approving, but curious of folks thoughts on questions I posed.

**Comment by CodeBlanch** [comment]
@alanwest 

> If we consider the default (ParseStateValues = false) as a bad default, we could change the default to make it true. To avoid breaking existing users, the SDK can continue to populate State as well. That way, exporters need to only deal with the parse state values, and if any exporters were previously relying on the State, they will be unaffected as well...

I liked this too initially. But then looking at some code, for example:

https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/1fffa1e265019bf03a7d1361fe46c8c2fb88dbba/src/OpenTelemetry.Exporter.Geneva/GenevaLogExporter.cs#L201

If we switched the behavior so we always set state, anyone who has their ordering as "check state, then check StateValues" we might be breaking those exporters if we do that.

> Another alternative may be to make the setting bool? ParseStateValues. Default is null - i.e., unspecified. Makes it clear that the actual default depends on the exporter you've configured.

I like that. Might still be breaking. We could `[Obsolete]` `ParseStateValues` and add `bool DisableState { get; set; }` maybe? If user sets `DisableState = true` we know they truly don't want anything. If `DisableState = false` we treat that as `ParseStateValues = true` today?

**Comment by alanwest** [comment]
> I like that. Might still be breaking. We could [Obsolete] ParseStateValues and add bool DisableState { get; set; } maybe? If user sets DisableState = true we know they truly don't want anything. If DisableState = false we treat that as ParseStateValues = true today?

Yea, you're right it would definitely be a breaking change. It kinda seems like someone would need to be doing something kinda funky to be affected by the change, but obviously that's not a good enough reason for a breaking change 😄.

I'm not opposed to this idea of adding `Obsolete`. Feels a little dirty, but it's certainly something we've considered in other contexts - like the OTLP batch activity processor options. I think it might be a bit cleaner than an unexpected side affect of calling `AddOtlpExporter`.

**Comment by cijothomas** [comment]
Merging to continue fixing the rest of OTLP LogExporter issues.
The question of how to deal with state definitely is worth coming back to, once we have a fully working solution for end users.

**Pull request merged**